Fix Genserver bottleneck

cachemoi opened the pull request:
This client periodically polls the unleash service and stores the fetched feature flag settings into genserver state, calling the genserver again each time we want to fetch a feature flag value.

This means that access to this state is sequential, thereby creating a bottleneck

The solution implemented here is to use an [ETS table instead of the genserver state](https://elixirforum.com/t/why-use-ets-when-not-to-use-it/4326/8) in order to allow for concurrent reads.

sgerrand commented on lib/unleash/backup.ex:
```diff
@@ -0,0 +1,57 @@
+        backup_dir \\ Config.backup_dir(),
+        backup_file \\ Config.backup_file()
+      ) do
+    if not File.dir?(backup_dir) do
+      backup_dir
+      |> File.mkdir_p!()
+    end
```
The [non-bang version](https://hexdocs.pm/elixir/File.html#mkdir_p/1) of this function might be simpler to use as it follows the Unix approach to `mkdir -p`: blindly create the entire directory structure.
```suggestion
    :ok = File.mkdir_p((backup_dir)
```

cachemoi committed:
Create ETS table backed cache

cachemoi committed:
Use ETS cache to store features rather than Genserver state

Storing state in a genserver means that we access cached feature flag values sequentially. Using ETS tables allows us to use concurrent reads to access this data, thereby fixing a performance bottleneck

cachemoi committed:
Use no bang version of mkdir_p

mkdir_p will blindly create the entire dir structure

cachemoi committed:
Module comment correction

joladev commented on lib/unleash/repo.ex:
```diff
@@ -1,124 +1,128 @@
 
-  defp write_state(state) do
+  defp write_file_state(features) do
     if not File.dir?(Config.backup_dir()) do
       Config.backup_dir()
-      |> File.mkdir_p!()
+      |> File.mkdir_p()
     end
```
what failure behavior do you want here? With this change any failures will be ignored, but it will instead fail on the `File.write! a few lines down, and crash the same

cachemoi replied:
As per [Sasha's comment](https://github.com/duffelhq/unleashed_ex/pull/2#discussion_r752141253) this change was to blindly create the whole directory path rather than check for it first (although i missed that in my code)

I've added sasha's full suggestion, it should raise a matcherror upon failure

joladev commented on lib/unleash/repo.ex:
```diff
@@ -1,124 +1,128 @@
   # threshold.
-  def handle_info({:mojito_response, _ref, _message}, state) do
-    {:noreply, state}
+  def handle_info({:mojito_response, _ref, _message}, _state) do
+    {:noreply, nil}
```
you initialise state to `%Feature{}` but change it here and a few other places to `nil` instead of keeping the original state

I would recommend using `[]` as the original state and then preserving it across the module, it makes it easier to add state in the future. `[]` is sort of a default empty, I guess it technically uses less memory than `nil`

joladev replied:
ah if `%Features{}` is used then use that as the state across the board and don't overwrite it with `nil`

cachemoi replied:
I changed it to be assigned to `[]` on init and not to re-assign it to nil in the `handle_info` handler

joladev commented on lib/unleash/repo.ex:
```diff
@@ -1,124 +1,128 @@
   alias Unleash.Features
+  alias Unleash.Cache
 
   def init(%Features{} = features) do
```
in what scenario is `%Features{}` the argument for init?

cachemoi replied:
It's used to pre-populate the cache in tests (e.g [here](https://github.com/duffelhq/unleashed_ex/blob/main/test/unleash_test.exs#L144) or [here](https://github.com/duffelhq/unleashed_ex/blob/main/test/unleash/variant_test.exs#L17) or [here](https://github.com/duffelhq/unleashed_ex/blob/main/test/unleash/client_specification_test.exs#L29)

cachemoi committed:
Slightly improve Genserver state usage

Change Genserver state to be [] instead of nil, don't reassign it in the handle_info handler.

cachemoi committed:
Raise matcherror upon failing to create backup dir

Now that we use the no-bang version of mkdir_p, we still need to raise an error upon failure

joladev reviewed (approved):
LGTM, do you feel like the test suite has covered it enough or should it be taken for a test drive in staging?

Also give others a chance to review in case we missed something

ulissesalmeida commented on lib/unleash/cache.ex:
```diff
@@ -0,0 +1,78 @@
+    end)
+  end
+
+  defp upsert_feature(name, value, table_name) when is_binary(name) do
```
🎨 `upsert_feature` => `insert` ?

Not a big deal, it's calling `insert` under the hood

cachemoi replied:
`Cache.upsert_feature` seems a bit more descriptive to me than `Cache.insert`, I'll keep it as is if that's OK with you

ulissesalmeida replied:
Yeah, no issue at all

ulissesalmeida reviewed (approved):
LGTM

Just 🎨  about the commit messages.

In the commit messages and in the PR title we're leaking the JIRA ticket ID of the task.
Our plan is to merge this on the upstream open source repository, the JIRA ticket seems like a leak of how Duffel works, not necessarily relevant to the upstream project. I would maybe remove the JIRA id from the commits before opening a PR for the GitLab project.

cachemoi commented:
> In the commit messages and in the PR title we're leaking the JIRA ticket ID of the task.

I've purged the ticket IDs +  🎨 the commit messages a little

> Should it be taken for a test drive in staging?

Since it's on the critical path for a lot of things I'll hold off on using it in prod until it's been confirmed to work in staging
